frontend/catalyst/cuda/primitives/__init__.py

```diff
@@ -477,6 +477,27 @@ def cudaq_spin_abs(target, kind):  # pragma: nocover
     return AbsCudaSpinOperator()
 
 
+cudaq_for_p = jax.core.Primitive("cudaq_for")
+
+
+def cudaq_for(kernel, start, stop, function):
+    """Convenience wrapper around the primitive."""
+    # what exactly does it do
+    return cudaq_for_p.bind(kernel, start, stop, function)
+
+
+@cudaq_for_p.def_impl
+def cudaq_for_impl(kernel, start, stop, function):
+    """Concrete implementation."""
+    return kernel.for_loop(start, stop, function)
+
+
+@cudaq_for_p.def_abstract_eval
+def cudaq_for_abs(start, stop, function):  # pragma: nocover
+    """Abstract for loop."""
+    return AbsCudaValue()
```

Comment:
I am unsure what a for loop should return during abstract evaluation, please advise

Reply:
Hi @zzzDavid,

Looking at the documentation of CUDA Quantum's API for [`for_loop`](https://nvidia.github.io/cuda-quantum/latest/api/languages/python_api.html#cudaq.Kernel.for_loop), it looks like it doesn't return anything:

```python
def foo(index: int):
    """A function that will be applied to `kernel` in a for loop."""
    kernel.cx(qubits[index], qubits[index+1])

# Create a for loop in `kernel`, providing a concrete number
# of iterations to run (4).
kernel.for_loop(start=0, stop=4, function=foo)
```

With JAX, the way we've implemented primitives that won't return anything is:

* Set the primitive to return multiple returns
* In the abstract evaluation simple return an empty tuple

This means that:
* If the Catalyst for-loop returns anything, then it won't be able to be converted into a CUDA for-loop.

Reply:
Thanks! I've marked it as returning multiple values and let it return empty tuples.